File: web/game.js
/* ---------- RENDERING FUNCTIONS ---------- */
function renderGame(state) {
    renderPlayers(state);
    renderStacks(state.stacks);
    renderSelf(state.self, state.self_conn);
    renderOthers(state.others);
}

/* ---------- Players ---------- */
function renderPlayers(state) {
    const el = document.getElementById("players-indicator");
    el.innerHTML = "";

    state.players.forEach(p => {
        const dot = document.createElement("div");
        dot.className = "player-dot";

        if (p === state.self_conn) dot.classList.add("self");
        if (p === state.waiting_for_stack) dot.classList.add("waiting");

        el.appendChild(dot);
    });
}

/* ---------- Stacks ---------- */
function renderStacks(stacks) {
    const el = document.getElementById("stacks");
    el.innerHTML = "";

    stacks.forEach((row, rowIndex) => {
        const rowEl = document.createElement("div");
        rowEl.className = "stack";

        row.forEach(card => {
            const cardEl = createCard(card);
            rowEl.appendChild(cardEl);
        });

        // Make stack clickable if the server expects a stack choice
        rowEl.onclick = () => {
            if (waitingForStack) {
                ws.send(JSON.stringify({
                    type: "choose_stack",
                    token: token,
                    index: rowIndex
                }));
                waitingForStack = false; // prevent double-click
            }
        };

        el.appendChild(rowEl);
    });
}

function sumCardPoints(cards) {
    return cards.reduce((sum, c) => sum + c.value, 0);
}

/* ---------- Self ---------- */
function renderSelf(self, self_conn) {
    const handEl = document.getElementById("hand");
    const playedEl = document.getElementById("played-card");
    const pointsEl = document.getElementById("self-points");

    playedEl.dataset.playerId = self_conn;

    handEl.innerHTML = "";
    playedEl.innerHTML = "";

    const total = self.points + sumCardPoints(self.card_points);
    pointsEl.textContent = `You (${self.name}): ${total} pts`;

    self.hand
     		.map((card, originalIndex) => ({card, originalIndex}))
    		.sort((a, b) => a.card.face - b.card.face)
    		.forEach(({card, originalIndex}) => {
        const cardEl = createCard(card);
        cardEl.onclick = () => {
            if (!self.played_card && !waitingForStack) {
                ws.send(JSON.stringify({
                    type: "play_card",
                    token,
                    index: originalIndex
                }));
            }
        };
        handEl.appendChild(cardEl);
    });

    if (self.played_card) {
        const card = createCard(self.played_card);
        card.classList.add("played");
        playedEl.appendChild(card);
    }
}

/* ---------- Others ---------- */
function renderOthers(others) {
    const container = document.getElementById("others-played");
    container.innerHTML = "";

    Object.entries(others).forEach(([player, data]) => {
        const wrapper = document.createElement("div");
        wrapper.className = "played-wrapper";
        wrapper.dataset.playerId = player;

        const total = data.points + sumCardPoints(data.card_points);

        const label = document.createElement("div");
        label.className = "played-label";
        label.textContent = `${data.name} (${total} pts)`;
        wrapper.appendChild(label);

        if (data.played_card) {
            const card = createCard(data.played_card);
            card.classList.add("played");
            wrapper.appendChild(card);
        }

        container.appendChild(wrapper);
    });
}

/* ---------- Card ---------- */
function createCard(card) {
	  const el = document.createElement("div");
	  el.className = `card value-${card.value}`;

	  el.innerHTML = `
		    <div class="face">${card.face}</div>
		    <div class="value">${card.value}</div>
	  `;
	  return el;
}

/* ---------- Animations ---------- */
function stackAddElement(i) {
  const stacks = document.getElementById("stacks");
  const row = stacks.children[i]; // the stack row
  const ph = document.createElement("div");
  ph.className = "card placeholder";
  row.appendChild(ph);
  return ph;
}

function flyCard(cardEl, fromEl, toEl, { duration = 300 } = {}) {
  const from = fromEl.getBoundingClientRect();
  const to = toEl.getBoundingClientRect();

  const clone = cardEl.cloneNode(true);
  clone.classList.add("flying");
  document.body.appendChild(clone);

  clone.style.transitionDuration = `${duration}ms`;

  // start position
  clone.style.left = `${from.left}px`;
  clone.style.top = `${from.top}px`;

  // force layout
  clone.getBoundingClientRect();

  const dx = to.left - from.left;
  const dy = to.top - from.top;

  return new Promise((resolve) => {
    let finished = false;

    const finish = () => {
      if (finished) return;
      finished = true;
      clone.remove();
      resolve();
    };

    clone.addEventListener("transitionend", finish, { once: true });

    // animate
    clone.style.transform = `translate(${dx}px, ${dy}px)`;

    // fallback in case transitionend doesn't fire
    setTimeout(finish, duration + 50);
  });
}

function createQueue() {
  let chain = Promise.resolve();

  return {
    add(task) {
      // task can be sync or async
      chain = chain.then(() => task()).catch(console.error);
      return chain; // optional: lets you await "until everything so far is done"
    }
  };
}

const q = createQueue();

function processAnimations(animations) {
  for (const anim of animations) {
    q.add(() => {
      if (anim.action === "play_card") {
      	const playerEl = document.querySelector(`[data-player-id="${anim.player}"]`);
        if (!playerEl) return;
        const start = playerEl.querySelector(".played");
        if (!start) return;

        const to = stackAddElement(anim.stack);
        return flyCard(start, start, to, { duration: 2000 });
      }

      if (anim.action === "pick_stack") {
     		const playerEl = document.querySelector(`[data-player-id="${anim.player}"]`);
        if (!playerEl) return;
        const stacksEl = document.getElementById("stacks");
        const stackEl = stacksEl.children[anim.stack];
        if (!stackEl) return;

        // animate cards one-by-one
        let p = Promise.resolve();
        for (const cardEl of stackEl.children) {
          p = p.then(() => flyCard(cardEl, cardEl, playerEl, { duration: 1000 }));
        }
        return p;
      }

      if (anim.action === "reveal_cards") {
        for (const [conn, card] of Object.entries(anim.revealed_cards)) {
        	const playerEl = document.querySelector(`[data-player-id="${conn}"]`);
          if (!playerEl) return;
          
          const cardEl = createCard(card);
          cardEl.classList.add("played");
          
          const old = playerEl.querySelector(".played");
          if (old) old.replaceWith(cardEl);
          else playerEl.appendChild(cardEl);
        }
      }
    });
  }
}

/* ---------- WEBSOCKET & SESSION ---------- */
const token = sessionStorage.getItem("token");
const addr  = sessionStorage.getItem("ws_addr");

if (!token || !addr) {
    alert("No saved session found, please login first.");
    window.location.href = "connect.html";
}

let waitingForStack = false;

const ws = new WebSocket(`ws://${addr}/ws`);

ws.onopen = () => {
    console.log("Connected to server, sending reconnect request…");
    ws.send(JSON.stringify({
        type: "reconnect",
        token: token
    }));
};

ws.onmessage = (e) => {
    let msg;
    try { msg = JSON.parse(e.data); } 
    catch { console.error("Invalid message:", e.data); return; }

    if (msg.type === "gamestate" || msg.type === "reconnect_ok") {
      waitingForStack = !!msg.state.waiting_for_stack;
      q.add(() => { renderGame(msg.state); });
    }
    
    if (msg.type === "animations") {
      processAnimations(msg.animations); // this enqueues animations
    }
};

ws.onerror = (e) => console.error("WebSocket error:", e);
ws.onclose = () => console.log("Connection closed by server.");

File: game.py
import random, json

class Game:
  def __init__(self):
    self.lobby_open = True
    self.waiting_for_card = set()
    self.waiting_for_stack = None
    self.players = {}
    self.deck = Deck([Card(i, 7 if i == 55 else
                    5 if i % 11 == 0 else
                    3 if i % 10 == 0 else
                    2 if i % 5 == 0 else
                    1) for i in range(1, 104+1)])
    self.stacks = tuple([] for _ in range(4))

  def get_player_infos(self, player):
   data = {
    "players": list(self.players.keys()),
    "waiting_for_stack": self.waiting_for_stack,
    "waiting_for_card": list(self.waiting_for_card),
    "stacks": self.stacks,
    "self_conn": player,
    "self": self.players[player],
    "others": {
      player_conn: {
        "name": player_obj.name,
        "played_card": Card("?", "?") 
                      if not self.waiting_for_stack and player_obj.played_card 
                      else player_obj.played_card,
        "points": player_obj.points,
        "card_points": player_obj.card_points
      } for player_conn, player_obj in self.players.items() if player_conn != player
    }
   }
   return data

  def add_player(self, player, name):
    if self.lobby_open and len(self.players) < 10:
      self.players[player] = Player(name)
      return True
    return False

  def start(self):
    self.lobby_open = False
    self._reset_round()
    self._all_waiting_for_card()

  def play_card(self, player, index):
    if player in self.waiting_for_card:
      if self.players[player].play_card(index):
        self.waiting_for_card.remove(player)
      actions = self._check_everyone_played_card()
      return actions

  def choose_stack(self, player, index):
    if player == self.waiting_for_stack and 0 <= index and index < 4:
      player = self.players[player]
      correct_stack = self.stacks[index]
      player.card_points.extend(correct_stack)
      correct_stack.clear()
      correct_stack.append(player.played_card)
      
      player.played_card = None
      self.waiting_for_stack = None

      res, actions = self._play_on_stacks()
      return actions

  def _play_on_stacks(self):
    actions = []
    player_conns = sorted((conn for conn, player in self.players.items() if player.played_card is not None), 
                    key=lambda conn: self.players[conn].played_card.face)
    for conn in player_conns:
      player = self.players[conn]
      correct_index, correct_stack = max(((i, s) for i, s in enumerate(self.stacks) if s[-1].face < player.played_card.face),
                  key=lambda x: x[1][-1].face,
                  default=(None, None))
      if not correct_stack:
        self.waiting_for_stack = conn
        return False, actions
      if len(correct_stack) >= 5:
        player.card_points.extend(correct_stack)
        correct_stack.clear()
        actions.append({"action": "pick_stack", "stack": correct_index, "player": conn})
      correct_stack.append(player.played_card)
      actions.append({"action": "play_card", "stack": correct_index, "player": conn})
      player.played_card = None
    self._all_waiting_for_card()
    self._check_end_of_round()
    return True, actions

  def _check_end_of_round(self):
    if len(next(iter(self.players.values())).hand) == 0:
      self._reset_round()

  def _check_everyone_played_card(self):
    if len(self.waiting_for_card) == 0:
      reveal_action = {"action": "reveal_cards", "revealed_cards": {conn: player.played_card for conn, player in self.players.items()}}
      res, actions = self._play_on_stacks()
      return [reveal_action] + actions

  def _all_waiting_for_card(self):
    self.waiting_for_card.update(self.players.keys())

  def _reset_round(self):
    for stack in self.stacks:
      self.deck.return_cards(stack)
      stack.clear()
    for player in self.players.values():
      player.count_cards()
      self.deck.return_cards(player.return_cards())
      
    self.deck.shuffle()
    
    for stack in self.stacks:
      cards = self.deck.draw(1)
      stack.extend(cards)
    for player in self.players.values():
      cards = self.deck.draw(10)
      player.add_cards(cards)


class Card:
  def __init__(self, face, value):
    self.face = face
    self.value = value


class Deck:
  def __init__(self, cards):
    self.cards = cards

  def shuffle(self):
    random.shuffle(self.cards)

  def draw(self, amount):
    cards = []
    for _ in range(amount):
      cards.append(self.cards.pop())
    return cards

  def return_cards(self, cards):
    self.cards.extend(cards)


class Player:
  def __init__(self, name):
    self.name = name
    self.hand = []
    self.card_points = []
    self.points = 0
    self.played_card = None

  def add_cards(self, cards):
    self.hand.extend(cards)

  def count_cards(self):
    self.points += sum((card.value for card in self.card_points))

  def play_card(self, i):
    if 0 <= i and i < len(self.hand) and not self.played_card: 
      self.played_card = self.hand.pop(i)
      return True
    return False

  def return_cards(self):
    cards = []
    cards.extend(self.hand)
    self.hand.clear()
    cards.extend(self.card_points)
    self.card_points.clear()
    if self.played_card:
      cards.append(self.played_card)
      self.played_card = None
      print("Not Supposed To Happen")
    return cards
